kernel/time/src/lib.rs

```diff
@@ -0,0 +1,322 @@
+    }
+}
+
+/// A clock frequency.
+#[derive(Clone, Copy, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub struct Frequency(u64);
+
+impl Frequency {
+    /// Creates a new frequency with the specified hertz.
+    pub fn new(frequency: u64) -> Self {
+        Self(frequency)
+    }
+}
+
+impl From<Frequency> for u64 {
+    /// Returns the frequency in hertz.
+    fn from(f: Frequency) -> Self {
+        f.0
+    }
+}
+
+impl From<u64> for Frequency {
+    /// Creates a new frequency with the specified hertz.
+    fn from(f: u64) -> Self {
+        Self(f)
+    }
+}
+
+/// Register the clock source that can be used to sleep when interrupts are
+/// disabled.
+///
+/// The current early sleeper will only be overwritten by `T` if `frequency` is
+/// larger than the frequency of the current early sleeper.
+///
+/// Returns whether the early sleeper was overwritten.
+pub fn register_early_sleeper<T>(frequency: Frequency) -> bool
+where
+    T: EarlySleeper,
+{
+    let old_frequency = EARLY_SLEEPER_FREQUENCY.load(Ordering::SeqCst);
+    if frequency > Frequency::new(old_frequency) {
+        EARLY_SLEEP_FUNCTION.store(T::sleep);
+        EARLY_SLEEPER_FREQUENCY.store(frequency.into(), Ordering::SeqCst);
+        true
+    } else {
+        false
+    }
+}
+
+/// Wait for the given `duration`.
+///
+/// This function spins the current task rather than sleeping it and so, when
+/// possible, the `sleep` crate should be used.
+///
+/// However, unlike the `sleep` crate, this function doesn't rely on interrupts,
+/// and can be used prior to the scheduler being initiated.
+pub fn early_sleep(duration: Duration) {
+    let f = EARLY_SLEEP_FUNCTION.load();
+    f(duration)
+}
+
+/// Register a clock source.
+///
+/// The provided source will overwrite the previous source only if `frequency`
+/// is larger than that of the previous source.
+///
+/// Returns whether the previous source was overwritten.
+pub fn register_clock_source<T>(frequency: Frequency) -> bool
+where
+    T: ClockSource,
+{
+    let old_frequency = Frequency::new(T::ClockType::frequency_atomic().load(Ordering::SeqCst));
+    if frequency > old_frequency {
+        let now_fn = T::ClockType::now_fn();
+        now_fn.store(T::now);
+
+        T::ClockType::store_unit_to_duration_func(T::unit_to_duration);
+        T::ClockType::store_duration_to_unit_func(T::duration_to_unit);
+
+        T::ClockType::frequency_atomic().store(frequency.into(), Ordering::SeqCst);
+
+        true
+    } else {
+        false
+    }
+}
+
+/// Returns the current time.
+///
+/// Monotonic clocks return an [`Instant`] whereas wall time clocks return a
+/// [`Duration`] signifying the time since 12:00am January 1st 1970 (i.e. Unix
+/// time).
+///
+/// # Panics
+///
+/// This function will panic if called prior to registering a clock using
+/// [`register_clock_source`]. [`register_clock_source`] must return [`Ok`] and
+/// the [`ClockType`] of the registered clock must be the same as `T`.
```

Comment:
i was just about to merge this when I saw this doc comment. Is this still accurate? I don't see any panics possible in this code path, so it's likely outdated.